File: finalyearproject/repos/dkd-platform-libs/sdk/python/src/dkd_platform/outbox.py
# HAND-AUTHORED platform primitive (NOT dkdgen-generated).
# PL-02 — transactional outbox + publisher relay (SA-CONV-QUARTET, EF §21.1 / EF-EVT-6).
"""Transactional outbox: the event row is inserted with the aggregate write, in the caller's tx.

Canonical schema (the SDK standard, already live in dkd-custody-ledger):

    outbox(id BIGSERIAL PK, event_id TEXT NOT NULL UNIQUE, topic TEXT, key TEXT, payload JSONB,
           occurred_at_ms BIGINT, created_at TIMESTAMPTZ default now(), published_at TIMESTAMPTZ NULL)
    CREATE INDEX outbox_unpublished_idx ON outbox (id) WHERE published_at IS NULL;

`enqueue` is idempotent on `event_id` (ON CONFLICT DO NOTHING) so a retried command never
double-emits. `OutboxRelay` drains unpublished rows in id order and stamps `published_at`; it
injects the `traceparent` (stub-safe: pass through when absent), `event_id` and `producer_context`
headers the fleet R6 convention requires.
"""
from __future__ import annotations

import json
from dataclasses import dataclass
from typing import Any

from .dbexec import DbExecutor

# `key` is a non-reserved word in Postgres but reads as a keyword — always quote it.
_INSERT = (
    'INSERT INTO outbox(event_id, topic, "key", payload, occurred_at_ms) '
    "VALUES (%s, %s, %s, %s::jsonb, %s) "
    "ON CONFLICT (event_id) DO NOTHING"
)
_SELECT_UNPUBLISHED = (
    'SELECT id, event_id, topic, "key", payload::text, occurred_at_ms FROM outbox '
    "WHERE published_at IS NULL ORDER BY id LIMIT %s"
)
_MARK_PUBLISHED = "UPDATE outbox SET published_at = now() WHERE id = ANY(%s)"

_MAX_BATCH = 500


def _json_param(payload: Any) -> str:
    """JSONB bind value. Dicts/lists are canonicalised; a pre-serialized str passes through."""
    if isinstance(payload, str):
        return payload
    return json.dumps(payload, sort_keys=True, separators=(",", ":"))


@dataclass(frozen=True)
class OutboxRecord:
    """The enqueue input — `{eventId, topic, key, payload, occurredAtMs}` in Python casing."""

    event_id: str
    topic: str
    key: str
    payload: Any
    occurred_at_ms: int


@dataclass(frozen=True)
class UnpublishedRow:
    id: int
    event_id: str
    topic: str
    key: str
    payload: str
    occurred_at_ms: int


class Outbox:
    """Write side — called inside the aggregate's transaction."""

    @staticmethod
    def enqueue(tx: DbExecutor, record: OutboxRecord) -> bool:
        """Insert one outbox row atomically with the caller's aggregate write.

        Returns True when the row was inserted, False when `event_id` already existed
        (ON CONFLICT DO NOTHING — a duplicate command is a silent no-op, never a double-emit).
        """
        cur = tx.execute(
            _INSERT,
            (
                record.event_id,
                record.topic,
                record.key,
                _json_param(record.payload),
                record.occurred_at_ms,
            ),
        )
        return cur.rowcount == 1


class OutboxRelay:
    """Publisher-loop side — reads unpublished rows and marks them published."""

    @staticmethod
    def fetch_unpublished(db: DbExecutor, limit: int) -> list[UnpublishedRow]:
        cur = db.execute(_SELECT_UNPUBLISHED, (max(1, min(limit, _MAX_BATCH)),))
        return [
            UnpublishedRow(
                id=r[0], event_id=r[1], topic=r[2], key=r[3], payload=r[4], occurred_at_ms=r[5]
            )
            for r in cur.fetchall()
        ]

    @staticmethod
    def mark_published(db: DbExecutor, ids: list[int]) -> int:
        if not ids:
            return 0
        cur = db.execute(_MARK_PUBLISHED, (list(ids),))
        return cur.rowcount

    @staticmethod
    def headers(
        row: UnpublishedRow, producer_context: str, traceparent: str | None = None
    ) -> list[tuple[str, bytes]]:
        """Kafka headers for a published row. `event_id` + `producer_context` are always present;
        `traceparent` is passed through only when the caller carries one (stub-safe when None)."""
        hdrs = [
            ("event_id", row.event_id.encode()),
            ("producer_context", producer_context.encode()),
        ]
        if traceparent:
            hdrs.append(("traceparent", traceparent.encode()))
        return hdrs

File: finalyearproject/repos/dkd-platform-libs/sdk/csharp/Dkd.Platform/Outbox.cs
// HAND-AUTHORED platform primitive (NOT dkdgen-generated).
// Transactional-outbox half of the effectively-once quartet (outbox + inbox + DLQ + idempotency).
// Canon: EF §21.1 / EF-EVT-6 / SA-CONV-QUARTET (outbox+inbox = effectively-once),
// SA-MSG-09/10 (DLQ + per-aggregate-key park-and-freeze). One of five byte-compatible SDK
// implementations (Go/Java/Node/Python/C#) exposing the SAME contract, so no service hand-rolls
// these helpers (PL-02). Matches the semantics of dkd-custody-ledger/internal/outbox/relay.go.

namespace Dkd.Platform;

/// <summary>
/// Minimal, driver-agnostic DB-execution seam. The service supplies the concrete transaction or
/// connection handle (Npgsql / Dapper / EF Core...) as a thin adapter; the SDK helpers never bind a
/// driver so they stay UNIT-TESTABLE without a live database. Positional <paramref name="args"/> map
/// to <c>$1..$N</c> placeholders in <paramref name="sql"/> (Npgsql/pg style).
/// </summary>
public interface IDbExecutor
{
    /// <summary>Run a non-query statement; returns the number of rows affected.</summary>
    int Execute(string sql, params object?[] args);

    /// <summary>Run a query; returns rows as ordered column-value lists (in SELECT order).</summary>
    IReadOnlyList<IReadOnlyList<object?>> Query(string sql, params object?[] args);
}

/// <summary>One unpublished outbox row drained by the relay loop.</summary>
public sealed record OutboxRecord(
    long Id, string EventId, string Topic, string Key, string Payload, long OccurredAtMs);

/// <summary>
/// Transactional outbox writer. <see cref="Enqueue"/> inserts the event row using the CALLER-PROVIDED
/// transaction handle so the aggregate mutation and the event row commit atomically (SA-CONV-QUARTET).
/// Canonical SDK schema (already used by custody):
/// <code>
/// outbox(id BIGSERIAL PK, event_id TEXT NOT NULL UNIQUE, topic TEXT, key TEXT, payload JSONB,
///        occurred_at_ms BIGINT, created_at TIMESTAMPTZ default now(), published_at TIMESTAMPTZ NULL)
/// -- partial index outbox_unpublished_idx ON (id) WHERE published_at IS NULL
/// </code>
/// </summary>
public static class Outbox
{
    private const string InsertSql =
        "INSERT INTO outbox (event_id, topic, key, payload, occurred_at_ms) " +
        "VALUES ($1,$2,$3,$4::jsonb,$5) ON CONFLICT (event_id) DO NOTHING";

    /// <summary>
    /// Insert one outbox row inside the caller's transaction. Idempotent: a duplicate
    /// <paramref name="eventId"/> is a no-op (<c>ON CONFLICT (event_id) DO NOTHING</c>).
    /// Returns <c>true</c> when a new row was written, <c>false</c> on a duplicate.
    /// </summary>
    public static bool Enqueue(IDbExecutor tx, string eventId, string topic, string key,
        string payload, long occurredAtMs)
    {
        ArgumentNullException.ThrowIfNull(tx);
        Require(eventId, nameof(eventId));
        Require(topic, nameof(topic));
        return tx.Execute(InsertSql, eventId, topic, key, payload, occurredAtMs) == 1;
    }

    internal static void Require(string value, string name)
    {
        if (string.IsNullOrEmpty(value))
        {
            throw new ArgumentException($"outbox: {name} is required", name);
        }
    }
}

/// <summary>
/// Publisher-loop side of the outbox: drains committed rows and marks them published only after the
/// broker acks (at-least-once; consumers dedup via the inbox on <c>event_id</c>). The relay stamps
/// each produced record with <c>event_id</c> + <c>producer_context</c> headers plus a W3C
/// <c>traceparent</c> (stub-safe: passed through when present, never fabricated).
/// </summary>
public static class OutboxRelay
{
    private const int MaxBatch = 500;

    private const string FetchSql =
        "SELECT id, event_id, topic, key, payload::text, occurred_at_ms " +
        "FROM outbox WHERE published_at IS NULL ORDER BY id LIMIT $1";

    /// <summary>Fetch up to <paramref name="limit"/> (1..500) unpublished rows, oldest first.</summary>
    public static IReadOnlyList<OutboxRecord> FetchUnpublished(IDbExecutor db, int limit)
    {
        ArgumentNullException.ThrowIfNull(db);
        var capped = Math.Max(1, Math.Min(limit, MaxBatch));
        var rows = db.Query(FetchSql, capped);
        var result = new List<OutboxRecord>(rows.Count);
        foreach (var r in rows)
        {
            result.Add(new OutboxRecord(
                Convert.ToInt64(r[0]),
                (string)r[1]!,
                (string)r[2]!,
                (string)r[3]!,
                (string)r[4]!,
                Convert.ToInt64(r[5])));
        }

        return result;
    }

    /// <summary>Mark the given outbox ids published (after broker acks). No-op on an empty set.</summary>
    public static void MarkPublished(IDbExecutor db, IReadOnlyList<long> ids)
    {
        ArgumentNullException.ThrowIfNull(db);
        if (ids is null || ids.Count == 0)
        {
            return;
        }

        db.Execute("UPDATE outbox SET published_at = now() WHERE id = ANY($1)", ids.ToArray());
    }

    /// <summary>
    /// Build the wire headers the relay injects on every produced record: <c>event_id</c> +
    /// <c>producer_context</c> (dedup + provenance), plus a W3C <c>traceparent</c> (PL-05). The
    /// caller's traceparent is parsed/validated via <see cref="W3CTrace"/> and injected only when
    /// well-formed — a malformed value is dropped, and no trace is ever fabricated.
    /// </summary>
    public static IReadOnlyDictionary<string, string> Headers(OutboxRecord record,
        string producerContext, string? traceparent = null)
    {
        ArgumentNullException.ThrowIfNull(record);
        var headers = new Dictionary<string, string>
        {
            ["event_id"] = record.EventId,
            ["producer_context"] = producerContext,
        };
        if (W3CTrace.TryParse(traceparent, out var tp))
        {
            W3CTrace.Inject(headers, tp!);
        }

        return headers;
    }
}
